src/common/NetworkEventManager.ts

```diff
@@ -0,0 +1,101 @@
+import { Protocol } from 'devtools-protocol';
+import { HTTPRequest } from './HTTPRequest.js';
+
+export type QueuedEvents = {
+  responseReceived: Protocol.Network.ResponseReceivedEvent;
+  promise: Promise<void>;
+  resolver: () => void;
+  loadingFinished?: Protocol.Network.LoadingFinishedEvent;
+  loadingFailed?: Protocol.Network.LoadingFailedEvent;
+};
+
+export type RedirectInfoMap = Array<{
+  event: Protocol.Network.RequestWillBeSentEvent;
+  fetchRequestId?: string;
+}>;
+
+/**
+ * @internal
+ *
+ * Helper class to track network events by request ID
+ */
+export class NetworkEventManager {
+  /*
+   * There are four possible orders of events:
+   *  A. `_onRequestWillBeSent`
+   *  B. `_onRequestWillBeSent`, `_onRequestPaused`
+   *  C. `_onRequestPaused`, `_onRequestWillBeSent`
+   *  D. `_onRequestPaused`, `_onRequestWillBeSent`, `_onRequestPaused`,
+   *     `_onRequestWillBeSent`, `_onRequestPaused`, `_onRequestPaused`
+   *     (see crbug.com/1196004)
+   *
+   * For `_onRequest` we need the event from `_onRequestWillBeSent` and
+   * optionally the `interceptionId` from `_onRequestPaused`.
+   *
+   * If request interception is disabled, call `_onRequest` once per call to
+   * `_onRequestWillBeSent`.
+   * If request interception is enabled, call `_onRequest` once per call to
+   * `_onRequestPaused` (once per `interceptionId`).
+   *
+   * Events are stored to allow for subsequent events to call `_onRequest`.
+   *
+   * Note that (chains of) redirect requests have the same `requestId` (!) as
+   * the original request. We have to anticipate series of events like these:
+   *  A. `_onRequestWillBeSent`,
+   *     `_onRequestWillBeSent`, ...
+   *  B. `_onRequestWillBeSent`, `_onRequestPaused`,
+   *     `_onRequestWillBeSent`, `_onRequestPaused`, ...
+   *  C. `_onRequestWillBeSent`, `_onRequestPaused`,
+   *     `_onRequestPaused`, `_onRequestWillBeSent`, ...
+   *  D. `_onRequestPaused`, `_onRequestWillBeSent`,
+   *     `_onRequestPaused`, `_onRequestWillBeSent`, `_onRequestPaused`,
+   *     `_onRequestWillBeSent`, `_onRequestPaused`, `_onRequestPaused`, ...
+   *     (see crbug.com/1196004)
+   */
+  requestWillBeSent = new Map<
```

Comment:
since we are extracting a new class: would it make sense to provide a more granular API that the Network manager needs instead of exposing raw maps. E.g., `takeRequestExtraInfo(id): ExtraInfo|undefined; getRequestPausedEvent(id)` etc.

Reply:
This is a good thought, let me see what I can do.

Reply:
Ok @OrKoN I just tried your suggestion and it ends up pulling most of `NetworkManager`'s logic into it. I think for now this PR is better left as more of a raw data structure container. `NetworkManager` is basically meant to queue incoming CDP events, untangle them, then fire Puppeteer's own network events. If we introduce a more semantic API it pulls the 'untangling' logic into it.

Reply:
okay, let's land this and I will give it a try as well.